README.md
# CØSMOS Faucet

[![version](https://img.shields.io/github/v/release/okp4/cosmos-faucet?style=for-the-badge&logo=github)](https://github.com/okp4/cosmos-faucet/releases)
[![lint](https://img.shields.io/github/actions/workflow/status/okp4/cosmos-faucet/lint.yml?branch=main&label=lint&style=for-the-badge&logo=github)](https://github.com/okp4/cosmos-faucet/actions/workflows/lint.yml)
[![build](https://img.shields.io/github/actions/workflow/status/okp4/cosmos-faucet/build.yml?branch=main&label=build&style=for-the-badge&logo=github)](https://github.com/okp4/cosmos-faucet/actions/workflows/build.yml)
[![test](https://img.shields.io/github/actions/workflow/status/okp4/cosmos-faucet/test.yml?branch=main&label=test&style=for-the-badge&logo=github)](https://github.com/okp4/cosmos-faucet/actions/workflows/test.yml)
[![maintainability](https://img.shields.io/codeclimate/maintainability/okp4/cosmos-faucet?style=for-the-badge)](https://codeclimate.com/github/okp4/cosmos-faucet/maintainability)
[![coverage](https://img.shields.io/codecov/c/github/okp4/cosmos-faucet?style=for-the-badge&token=0VQHJDMY5B&logo=codecov)](https://codecov.io/gh/okp4/cosmos-faucet)
[![conventional commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg?style=for-the-badge&logo=conventionalcommits)](https://conventionalcommits.org)
[![contributor covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg?style=for-the-badge)](https://github.com/okp4/.github/blob/main/CODE_OF_CONDUCT.md)
[![license](https://img.shields.io/badge/License-BSD_3--Clause-blue.svg?style=for-the-badge)](https://opensource.org/licenses/BSD-3-Clause)

This faucet app allows anyone to request tokens for a [Cosmos](https://cosmos.network/) account address, either in command line or in service mode, with optional [ReCaptcha](https://www.google.com/recaptcha/about/) security.

The faucet app supports any Cosmos based blockchain and is intended to be configured on testnet networks.

## Install

```sh
go install github.com/okp4/cosmos-faucet@latest

cosmos-faucet --help
```

## Config

 Configuration can be passed as CLI argument, configuration file (`config.yml`) or by environment variable prefixed by `FAUCET` (i.e.: `FAUCET_MNEMONIC`).

```yml
grpc-address: 127.0.0.1:9090
mnemonic: "kiwi nuclear blast wet badge..."
chain-id: testnet-1
denom: uatom
prefix: cosmos
fee-amount: 0
amount-send: 1
memo: "Sent by økp4 faucet"
gas-limit: 200000
```

## Usage

### CLI

#### Send

```shell
Send tokens to a given address

Usage:
  cosmos-faucet send <address> [flags]

Flags:
  -h, --help   help for send

Global Flags:
      --amount-send int       Amount send value (default 1)
      --chain-id string       The network chain ID (default "localnet-okp4-1")
      --denom string          Token denom (default "know")
      --fee-amount int        Fee amount
      --gas-limit uint        Gas limit (default 200000)
      --grpc-address string   The grpc okp4 server url (default "127.0.0.1:9090")
      --memo string           The memo description (default "Sent by økp4 faucet")
      --mnemonic string
      --no-tls                No encryption with the GRPC endpoint
      --prefix string         Address prefix (default "okp4")
      --tls-skip-verify       Encryption with the GRPC endpoint but skip certificates verification
      --tx-timeout duration   Transaction timeout (default 5s)
```

#### Start

```shell
Start the GraphQL api

Usage:
  cosmos-faucet start [flags]

Flags:
      --address string              graphql api address (default ":8080")
      --batch-window duration       Batch temporal window, can be seen a the minimum duration between too transactions. (default 8s)
      --captcha                     enable captcha verification
      --captcha-min-score float     set Captcha min score (default 0.5)
      --captcha-secret string       set Captcha secret
      --captcha-verify-url string   set Captcha verify URL (default "https://www.google.com/recaptcha/api/siteverify")
      --health                      enable health endpoint
  -h, --help                        help for start
      --metrics                     enable metrics endpoint

Global Flags:
      --amount-send int       Amount send value (default 1)
      --chain-id string       The network chain ID (default "localnet-okp4-1")
      --denom string          Token denom (default "know")
      --fee-amount int        Fee amount
      --gas-limit uint        Gas limit (default 200000)
      --grpc-address string   The grpc okp4 server url (default "127.0.0.1:9090")
      --memo string           The memo description (default "Sent by økp4 faucet")
      --mnemonic string
      --no-tls                No encryption with the GRPC endpoint
      --prefix string         Address prefix (default "okp4")
      --tls-skip-verify       Encryption with the GRPC endpoint but skip certificates verification
      --tx-timeout duration   Transaction timeout (default 5s)
```

### GraphQL

Start GraphQL server with captcha verification for the send token mutation and subscription.

```shell
cosmos-faucet start --captcha --captcha-secret $CAPCTHA_SECRET
```

Access on playground and documentation at the root of server.

## Build

The project comes with a convenient `Makefile` which depends on [Docker](https://www.docker.com). Please verify that Docker is properly installed and if not, follow the instructions:

- for macOS: <https://docs.docker.com/docker-for-mac/install/>
- for Windows: <https://docs.docker.com/docker-for-windows/install/>
- for Linux: <https://docs.docker.com/engine/install/>

To build the app, invoke the following goal `build` of the `Makefile`:

```sh
make build
```

The app will be generated under the folder `target/dist`.

## :heart: Supporting this project & Contributing

A simple star in this repository is enough to make us happy!

But you're also welcome to contribute! We appreciate any help you're willing to give. Don't hesitate to open issues and/or submit pull requests.

Remember that this is the Cosmos Token Faucet we use at [OKP4](http://okp4.network). This is why we may have to refuse change requests simply because they do not comply with our internal requirements, and not because they are not relevant.


CHANGELOG.md
# CØSMOS Faucet

## [2.0.0](https://github.com/okp4/cosmos-faucet/compare/v1.1.0...v2.0.0) (2023-06-13)


### ⚠ BREAKING CHANGES

* **cosmos:** bump cosmos sdk version to 0.46.13

### Features

* **cosmos:** bump cosmos sdk version to 0.46.13 ([911a649](https://github.com/okp4/cosmos-faucet/commit/911a64934641ee05e346efaf76afad8095c8a793))

# [1.1.0](https://github.com/okp4/cosmos-faucet/compare/v1.0.0...v1.1.0) (2022-08-30)


### Bug Fixes

* **grpc:** properly stop grpc conn when actor stops ([1e4ced0](https://github.com/okp4/cosmos-faucet/commit/1e4ced0bd505ad411804308f421c286e3229329b))
* make it build again ([dbaa637](https://github.com/okp4/cosmos-faucet/commit/dbaa637bf0d90e70f169a279d4e8a78e12df392c))
* make linters happy ([363dc8f](https://github.com/okp4/cosmos-faucet/commit/363dc8ff3ce98b253c3c5f72721325bc348053b4))
* minor imprecisions in tx handler ([835f283](https://github.com/okp4/cosmos-faucet/commit/835f28309d9cd57ca9c77216419be6f8178ae4b6))
* **pool:** avoid sending empty transactions ([c1c4697](https://github.com/okp4/cosmos-faucet/commit/c1c46974fa8b780280990375b2b6bebc882bcb22))
* **pool:** close channels in only one goroutine ([297b0f4](https://github.com/okp4/cosmos-faucet/commit/297b0f4786bd57aaac82fceb8193665d76d15021))


### Features

* add warn log on non 0 tx code ([b377e5f](https://github.com/okp4/cosmos-faucet/commit/b377e5fb5d7c31cf136f2158e77adfba3d7e6dec))
* allow all origins ([d9ef2e8](https://github.com/okp4/cosmos-faucet/commit/d9ef2e81ebc7d127d8aa49e98458ce64656e0f75))
* allow send & subscribe symulnateously ([8fece85](https://github.com/okp4/cosmos-faucet/commit/8fece8550f17464b3de18968ecedb3b216438284))
* allow to set deadline on tx triggers ([bbf6451](https://github.com/okp4/cosmos-faucet/commit/bbf64515f1d3bb9ca5deda393ca93028a9876dd8))
* **api:** implements send graphql subscription ([dad1aae](https://github.com/okp4/cosmos-faucet/commit/dad1aae2970d11acb885b261b344cd92e25823bb))
* **api:** introduce the void scalar ([230de82](https://github.com/okp4/cosmos-faucet/commit/230de820e835a0649db1871e97ad8fd3871821ca))
* **cli:** add temporal batch window flag ([566831c](https://github.com/okp4/cosmos-faucet/commit/566831c26ae728172212a6080c7e91e40f2bf43c))
* **cli:** implement transaction timeout config ([973570c](https://github.com/okp4/cosmos-faucet/commit/973570ca2a220367ce0ca3fb8e02dde84d4392ac))
* implement an atomic message pool ([99634d8](https://github.com/okp4/cosmos-faucet/commit/99634d888f61fd800044a3d769b46b5a8b3bca07))
* implement message pool tx submitter option ([c881f68](https://github.com/okp4/cosmos-faucet/commit/c881f68d71d83f9be1b7b1d10553d0b9f63624d7))
* implements msg batching ([77955fe](https://github.com/okp4/cosmos-faucet/commit/77955fea69098cce82b624ff6ae6b05a20afe6cb))
* **metrics:** disable prometheus middleware ([8504e60](https://github.com/okp4/cosmos-faucet/commit/8504e60e7e25d1719c2a96253490956feed5c4fb))
* setup basic faucet trigger channel ([6ef0d4f](https://github.com/okp4/cosmos-faucet/commit/6ef0d4fd8de1b91e9ce51f526ec0a28e43628453))
* wire the faucet with the message pool ([ea3da76](https://github.com/okp4/cosmos-faucet/commit/ea3da76cf185a0814e19b4599d6c59e1a901b899))

# 1.0.0 (2022-05-31)


### Bug Fixes

* add method on graphql route ([ea5674e](https://github.com/okp4/cosmos-faucet/commit/ea5674e40d853ab68861f240cacbf105f2d1c478))
* change log type of health handler ([c460a98](https://github.com/okp4/cosmos-faucet/commit/c460a98937dd92954c886ec19bcb671229ee3af9))
* effectively set grpc tls parameters ([390170b](https://github.com/okp4/cosmos-faucet/commit/390170b1085219bc1e33cb94990280176165aaac))
* error handling on config file ([594941f](https://github.com/okp4/cosmos-faucet/commit/594941fed7d788b08752aa115f0869c25f1d15e5))
* handle grpc error ([7f35eb5](https://github.com/okp4/cosmos-faucet/commit/7f35eb528feca26a8608504f6ace55cc9bd0fbec))
* linter on github action ([ba32216](https://github.com/okp4/cosmos-faucet/commit/ba322164aaebbbb66d88e95d7b9ca01e4bf9d0db))
* make graphql linter happy ([c951771](https://github.com/okp4/cosmos-faucet/commit/c951771c40fdf50cd2147fe7122972c47c534829))
* metrics now correctly instantiated ([80ae76d](https://github.com/okp4/cosmos-faucet/commit/80ae76da69511d3024b05c98e02b3ed20a4cda0e))
* move server related flags to cmd/start.go ([9757c30](https://github.com/okp4/cosmos-faucet/commit/9757c3087211d22cd0684da9e77725b7d4a11023))
* remove unused send http api handler ([ea3c8d5](https://github.com/okp4/cosmos-faucet/commit/ea3c8d50f4e88fae9da8334da618da2487ffb3b1))
* removed unused param in newSendRequestHandler ([ae4c014](https://github.com/okp4/cosmos-faucet/commit/ae4c014f597782fb31d0ee55108fbcf928afb67c))
* retrive account on each transaction ([b9fbe59](https://github.com/okp4/cosmos-faucet/commit/b9fbe59e43f4ed26364c4e39a62e84bf79f76b75))
* some linter error ([23397b6](https://github.com/okp4/cosmos-faucet/commit/23397b6d17b711ea7071a2a0847721bb6c78e053))
* use request context for make transacation call ([a16d4ec](https://github.com/okp4/cosmos-faucet/commit/a16d4ec6c1b54e12e0d60353b0b65352c661a5c2))


### Features

* add an interface to expose HttpServer ([3810d8e](https://github.com/okp4/cosmos-faucet/commit/3810d8e919e32d52830a858a9afb1c0a62f7fdcd))
* add captcha secret flag to start cmd ([f0aa7fe](https://github.com/okp4/cosmos-faucet/commit/f0aa7fe560553a79f43a5d92137d92f2fa38fbd2))
* add flags for metrics and health endpoint ([1c9688f](https://github.com/okp4/cosmos-faucet/commit/1c9688f1398fa0108ce41d3d1336e58071a18def))
* add health check endpoint and base for prometheus metrics ([bb6e73a](https://github.com/okp4/cosmos-faucet/commit/bb6e73a256c31075cc089e57182c952d3681858c))
* add logs and error handling ([ea9d97e](https://github.com/okp4/cosmos-faucet/commit/ea9d97eb1a9cc0c0c133ff46f67739b7f07cb559))
* add query to fetch configuration of faucet ([75e7145](https://github.com/okp4/cosmos-faucet/commit/75e71451cee994e9baf9bf1f18aece8d861b739a))

cmd/root.go
package cmd

import (
	"crypto/tls"
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
)

const (
	defaultConfigFilename = "config"
	envPrefix             = "FAUCET"
)

var (
	mnemonic      string
	chainID       string
	grpcAddress   string
	denom         string
	prefix        string
	feeAmount     int64
	amountSend    int64
	memo          string
	gasLimit      uint64
	noTLS         bool
	tlsSkipVerify bool
	txTimeout     time.Duration
)

// NewRootCommand returns the root CLI command with persistent flag handling.
var rootCmd = &cobra.Command{
	Use:   "cosmos-faucet",
	Short: "A CØSMOS Faucet",
	Long:  "CØSMOS Faucet",
	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: cmd.OutOrStdout()})
		return ReadPersistentFlags(cmd)
	},
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main.main(). It only needs to happen once to the rootCmd.
func Execute() {
	rootCmd.PersistentFlags().StringVar(&mnemonic, FlagMnemonic, "", "")
	rootCmd.PersistentFlags().StringVar(&chainID, FlagChainID, "localnet-okp4-1", "The network chain ID")
	rootCmd.PersistentFlags().StringVar(&grpcAddress, FlagGrpcAddress, "127.0.0.1:9090", "The grpc okp4 server url")
	rootCmd.PersistentFlags().StringVar(&denom, FlagDenom, "know", "Token denom")
	rootCmd.PersistentFlags().StringVar(&prefix, FlagPrefix, "okp4", "Address prefix")
	rootCmd.PersistentFlags().Int64Var(&feeAmount, FlagFeeAmount, 0, "Fee amount")
	rootCmd.PersistentFlags().Int64Var(&amountSend, FlagAmountSend, 1, "Amount send value")
	rootCmd.PersistentFlags().StringVar(&memo, FlagMemo, "Sent by økp4 faucet", "The memo description")
	rootCmd.PersistentFlags().Uint64Var(&gasLimit, FlagGasLimit, 200000, "Gas limit")
	rootCmd.PersistentFlags().BoolVar(&noTLS, FlagNoTLS, false, "No encryption with the GRPC endpoint")
	rootCmd.PersistentFlags().BoolVar(&tlsSkipVerify,
		FlagTLSSkipVerify,
		false,
		"Encryption with the GRPC endpoint but skip certificates verification")
	rootCmd.PersistentFlags().DurationVar(&txTimeout, FlagTxTimeout, 5*time.Second, "Transaction timeout")

	err := rootCmd.Execute()
	if err != nil {
		os.Exit(1)
	}
}

func ReadPersistentFlags(cmd *cobra.Command) error {
	v := viper.New()

	v.SetConfigName(defaultConfigFilename)
	v.AddConfigPath(".")

	if err := v.ReadInConfig(); err != nil {
		// It's okay if there isn't a config file
		var configFileNotFound viper.ConfigFileNotFoundError
		if !errors.As(err, &configFileNotFound) {
			return err
		}
	}
	v.SetEnvPrefix(envPrefix)
	v.AutomaticEnv()
	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))

	cmd.Flags().VisitAll(func(f *pflag.Flag) {
		if !f.Changed && v.IsSet(f.Name) {
			val := v.Get(f.Name)
			var _ = cmd.Flags().Set(f.Name, fmt.Sprintf("%v", val))
		}
	})

	return nil
}

func getTransportCredentials() credentials.TransportCredentials {
	switch {
	case noTLS:
		return insecure.NewCredentials()
	case tlsSkipVerify:
		return credentials.NewTLS(&tls.Config{InsecureSkipVerify: true}) // #nosec G402 : skip lint since it's an optional flag
	default:
		return credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})
	}
}
